### DISCUSSION TITLE
How to convert cloudvolume's Skeleton object to navis.TreeNeuron or navis.TreeNeuronList

### DISCUSSION THREAD

**Discussion by liuyx599** [navis-org/navis#143] [General]
Hello, I'm just learning to use navis. in general, I can get a skeleton or a list of skeletons object by using cloudvolume as shown below:
```
    cv = cloudvolume.CloudVolume("precomputed://xxx")
    seg_ids = [1001, 1002, 1003]       # Assuming these are the actual segment_id's in cloudvolume
    skels = cv.skeleton.get(seg_ids)   # skels is a list of many Skeleton objects
    
    seg_id = 1004
    skel = cv.skeleton.get(seg_id)      # skel is a Skeleton object
 ```
Suppose I want to use some api in navis now, such as `navis.downsample_neuron`, etc., is there any good way to implement the object conversion, to **convert the Skeleton object obtained by cloudvolume directly to NeuronTree object supported by navis**, one way I can think of is. Download each skeleton in skels and save it as a swc file. Then read it with `navis.read_swc`, however this doesn't seem to be very efficient when analyzing against Pb level data

**Comment by schlegelp** [comment]
Hi Achilles,

What you are looking for is [`navis.patch_cloudvolume()`](https://navis.readthedocs.io/en/latest/source/generated/navis.patch_cloudvolume.html#navis.patch_cloudvolume). By running this function before you initialize the `CloudVolume` you can do `cv.skeleton.get_navis(1004)` and it returns the skeleton as navis `TreeNeuron`. See also [this](https://navis.readthedocs.io/en/latest/source/tutorials/cloudvolume_tut.html) short tutorial.

More concretely:
```python
navis.patch_cloudvolume()
cv = cloudvolume.CloudVolume("precomputed://xxx")
seg_ids = [1001, 1002, 1003]       # Assuming these are the actual segment_id's in cloudvolume
skels = cv.skeleton.get_navis(seg_ids)   # skels is a navis NeuronList with many skeletons

seg_id = 1004
skel = cv.skeleton.get_navis(seg_id)      # skel is a navis NeuronList with a single TreeNeuron
```

Let me know if you run into any issues.

**Comment by schlegelp** [comment]
PS: if you know the exact URL of the precomputed skeleton file(s), you can also use `navis.read_precomputed(url)` without needing to go through `cloudvolume`.

**Comment by liuyx599** [comment]
Thanks for your reply, as per your suggestion, I wrote and tried to run the following code (from H01 dataset):
```py
import navis
import cloudvolume as cv
navis.patch_cloudvolume()

vol_seg = cv.CloudVolume("precomputed://gs://h01-release/data/20210601/c3", cache=True, use_https=True)

seg_ids = [635743109, 2120664699]
skels = vol_seg.skeleton.get_navis(seg_ids)
```
Unfortunately there was an error,
```
INFO  : cloud-volume successfully patched! (navis)
100%|█████████████████████████████████████████████| 2/2 [00:00<00:00, 19.42it/s]
Shard Indices: 0it [00:00, ?it/s]
Shard Indices: 100%|██████████████████████████████| 2/2 [00:01<00:00,  1.13it/s]
Upload to Cache: 100%|████████████████████████████| 2/2 [00:00<00:00, 17.68it/s]
Minishard Indices: 0it [00:00, ?it/s]
Minishard Indices: 100%|██████████████████████████| 2/2 [00:01<00:00,  1.23it/s]
Upload to Cache: 100%|████████████████████████████| 2/2 [00:00<00:00, 17.54it/s]
Downloading Bundles: 100%|████████████████████████| 2/2 [00:01<00:00,  1.09it/s]
Decompressing: 100%|█████████████████████████████| 2/2 [00:00<00:00, 159.48it/s]
Upload to Cache: 100%|████████████████████████████| 2/2 [00:00<00:00, 17.69it/s]
Traceback (most recent call last):
  File "demo.py", line 25, in <module>
    skels = vol_seg.skeleton.get_navis(seg_ids)
  File "/src/anaconda3/envs/emseg/lib/python3.8/site-packages/navis/utils/cv.py", line 95, in wrapper
    for k, v in res.items():
AttributeError: 'tuple' object has no attribute 'items'
```

I single-step debugged into the `/navis/utils/cv.py` file, and I found that
``` py
for k, v in res.items(): 
# res is a tuple of cv.Skeleton objects with no k value instead of a dictionary
# just like ([Skeleton(segid=635743109...), Skeleton(segid=2120664699...)
    if isinstance(v, cv.Mesh):
        n = core.MeshNeuron(v, id=k, units='nm')
        neurons.append(n)
    elif isinstance(v, cv.Skeleton).
        swc_str = v.to_swc()
        n = io.read_swc(swc_str)
        n.id = k # n.id should be replaced by v.id here.
        n.units = 'nm'
        neurons.append(n)
```

My cloudvolume and navis's version:
```
navis.__version__
'1.5.0'
cv.__version__
'8.29.1'
```

**Comment by liuyx599** [comment]
I copied the above code for traversing res.item() into the main function at the very beginning with a slight modification and used vol_seg.skeleton.get(seg_ids) to read it instead of going through navis, which seems to give me a complete navis.NeuronList object at this point. I have a new question though, if I were to convert as follows, it also feels like it would take longer when there are a particularly large number of seg_ids. If I have run it once now, can I save the neuronList object below so that I can just read it directly when I want to use it next time.
```py
import navis
import cloudvolume as cv
navis.patch_cloudvolume()

vol_seg = cv.CloudVolume("precomputed://gs://h01-release/data/20210601/c3", cache=True, use_https=True)

seg_ids = [635743109, 2120664699]
skels = vol_seg.skeleton.get(seg_ids)

neurons = []
for skel in skels:
    if isinstance(skel, cv.Skeleton):
        swc_str = skel.to_swc()
        n = io.read_swc(swc_str)
        n.id = skel.id
        n.units = 'nm'
        neurons.append(n)

neuron_list = core.NeuronList(neurons)
```

**Comment by schlegelp** [comment]
Oh, that's strange. For me (on the same cloudvolume version) `vol.skeleton.get` always returns a list not a tuple of skeletons which is then correctly converted by `navis`.

In any event: the issue should hopefully be fixed with cd4be358ff9b0b7a6ec94abfea504917570bf937. Would you mind re-installing navis from Github and trying again?

```bash 
pip uninstall navis -y 
pip3 install git+https://github.com/navis-org/navis@master
```
